### PR TITLE
CMake: fix detection of hidapi

### PR DESCRIPTION
**Pull request by Be-ing**
Mixxx was building its vendored copy of hidapi even when the system had hidapi 0.10.1.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2605,9 +2605,8 @@ find_package(LibUSB)
 # USB HID controller support
 option(HID "USB HID controller support" ON)
 if(HID)
-  find_package(hidapi)
```
Comment (uklotzde) [inline CMakeLists.txt:2608]: 0.10.1 is the minimum required version, i.e. `0.10.1<` is what we actually want here. If a newer version breaks the build an upper bound could be introduced as a temporary workaround before finishing the upgrade.

Unfortunately, version ranges are only supported starting with CMake 3.19: https://cmake.org/cmake/help/latest/command/find_package.html

File: CMakeLists.txt [hunk 2]
```diff
@@ -2630,7 +2629,7 @@ if(HID)
     endif()
     target_link_libraries(mixxx-lib PRIVATE mixxx-hidapi)
   else()
-    message(STATUS "Linking libhidapi dynamically")
+    message(STATUS "Linking system libhidapi")
```
Comment (uklotzde) [inline CMakeLists.txt:2632]: This message now differs from all other external library "Linking ..." messages. Either change all or none.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (Be-ing) [reply CMakeLists.txt:2632]: Changing all is beyond the scope of this PR. This message is misleading. The system library may be a static library.

File: CMakeLists.txt (hunk 1 above)
Comment (Be-ing) [reply CMakeLists.txt:2608]: That is not valid:
```
CMake Error at CMakeLists.txt:2608 (find_package):
  find_package called with invalid argument "0.10.1<"
```

Comment (uklotzde) [reply CMakeLists.txt:2608]: Probably `0.10.1...<`
https://cmake.org/cmake/help/latest/command/find_package.html

Comment (Be-ing) [reply CMakeLists.txt:2608]: No, that does not work either:
```
CMake Error at CMakeLists.txt:2608 (find_package):
  find_package called with invalid argument "0.10.1...<"
```

I presume anything that would match `if(VERSION_GREATER_EQUAL)` would work, but the CMake documentation is not explicit about this. If you want to specify only one specific version, the documentation says to use EXACT.

Comment (uklotzde) [reply CMakeLists.txt:2608]: I have misunderstood the documentation. You can either specify a single (minimum?) version or both.

File: CMakeLists.txt (hunk 2 above)
Comment (uklotzde) [reply CMakeLists.txt:2632]: The let's leave the message as is. It causes more harm to introduce inconsistencies.

**Comment by uklotzde** [comment]
I suppose we are not able to resolve this issue in 2.3 without touching the macOS and Windows environments. Our include paths are wrong.

**Review by uklotzde — APPROVED** [review]
This went smoother than expected.

File: CMakeLists.txt [hunk 3]
```diff
@@ -2605,12 +2605,12 @@ find_package(LibUSB)
...
 option(HID "USB HID controller support" ON)
 if(HID)
-  find_package(hidapi)
-  # hidapi_VERSION is only available starting with 0.10.0
-  if(NOT hidapi_FOUND OR NOT hidapi_VERSION OR hidapi_VERSION VERSION_LESS 0.10.1)
+  find_package(hidapi 0.10.1)
+  if(NOT hidapi_FOUND)
     message(STATUS "Linking internal libhidapi statically")
     add_library(mixxx-hidapi STATIC EXCLUDE_FROM_ALL)
     target_include_directories(mixxx-hidapi SYSTEM PUBLIC lib/hidapi/hidapi)
```
Comment (uklotzde) [inline CMakeLists.txt:2612]: This should be PRIVATE, not PUBLIC.

File: CMakeLists.txt [hunk 4]
```diff
@@ -2605,12 +2605,12 @@ find_package(LibUSB)
 # USB HID controller support
 option(HID "USB HID controller support" ON)
 if(HID)
-  find_package(hidapi)
-  # hidapi_VERSION is only available starting with 0.10.0
-  if(NOT hidapi_FOUND OR NOT hidapi_VERSION OR hidapi_VERSION VERSION_LESS 0.10.1)
+  find_package(hidapi 0.10.1)
```
Comment (daschuer) [inline CMakeLists.txt:2608]: What was wrong with this expression above?
Find_package(hidapi 0.10.1) succeeds on Ubuntu Bionic, because it is no version number set.

**Review by daschuer — CHANGES_REQUESTED** [review]
I think the version detection in findhidapi.cmake requires an update to fix the issue.

File: CMakeLists.txt (hunk 4 above)
Comment (Be-ing) [reply CMakeLists.txt:2608]: Oof, why does CMake let this succeed...

Comment (Be-ing) [reply CMakeLists.txt:2608]: I fixed this by adding `hidapi_VERSION` to the `find_package_handle_standard_args` call.

Comment (Be-ing) [reply CMakeLists.txt:2608]: I think the error makes it clear what is happening:
```
-- Could NOT find hidapi (missing: hidapi_VERSION) (Required is at least version "0.10.1")
-- Linking internal libhidapi statically
```

**Comment by daschuer** [comment]
Now it prints: 
```
-- Could NOT find hidapi (missing: hidapi_VERSION) 
```
which is misleading. There is a ```Version: 0.9.0``` entry in the pc file. Cant we read that?

**Comment by Be-ing** [comment]
I think it is good enough how it is. It prints the required version already.

File: cmake/modules/Findhidapi.cmake [hunk 5]
```diff
@@ -62,28 +61,32 @@ find_library(hidapi_LIBRARY
...
-  hidapi
-  DEFAULT_MSG
-  hidapi_LIBRARY
-  hidapi_INCLUDE_DIR
-)
-
 # Version detection
-if (EXISTS "${hidapi_INCLUDE_DIR}/hidapi.h")
-    file(READ "${hidapi_INCLUDE_DIR}/hidapi.h" hidapi_H_CONTENTS)
+if (EXISTS "${hidapi_INCLUDE_DIR}/hidapi/hidapi.h")
```
Comment (daschuer) [inline cmake/modules/Findhidapi.cmake:65]: We are not sure if the header is in the /hidapi sub-directory. 

I am sure the original code was working when it has been written by @uklotzde here:
https://github.com/mixxxdj/mixxx/pull/3730
 
So for a reliable version detection we need to use the 
```Cflags: -I${includedir}/hidapi```
Info.

Comment (Be-ing) [reply cmake/modules/Findhidapi.cmake:65]: > We are not sure if the header is in the /hidapi sub-directory.

Yes we are. This is easily verifiable by looking up the list of files installed by packages. Here is the libhidapi-dev package from Ubuntu 18.04:
https://packages.ubuntu.com/bionic/amd64/libhidapi-dev/filelist

It installs /usr/include/hidapi/hidapi.h.

I am not sure this code ever worked before.

**Comment by daschuer** [comment]
Why don't we just use PC_hidapi_VERSION?

**Comment by daschuer** [comment]
> I think it is good enough how it is. It prints the required version already.

No, sorry. "Could NOT find hidapi" is wrong and misleading.

File: cmake/modules/Findhidapi.cmake (hunk 5 above)
Comment (daschuer) [reply cmake/modules/Findhidapi.cmake:65]: I can confirm that for Ubuntu Groovy. Maybe @uklotzde can explain why the code was written like that. Is it different with fedora?

**Comment by Be-ing** [comment]
I don't really care about perfect error messages for outdated operating systems. It still works and it gives a hint what is wrong:

```
Required is at least version "0.10.1"
```

File: cmake/modules/Findhidapi.cmake (hunk 5 above)
Comment (Be-ing) [reply cmake/modules/Findhidapi.cmake:65]: No, it is the same on Fedora:
```
❯ rpm -ql hidapi-devel
/usr/include/hidapi
/usr/include/hidapi/hidapi.h
/usr/lib64/libhidapi-hidraw.so
/usr/lib64/libhidapi-libusb.so
/usr/lib64/pkgconfig/hidapi-hidraw.pc
/usr/lib64/pkgconfig/hidapi-libusb.pc
```

**Comment by uklotzde** [comment]
@daschuer If you need better error messages for an issue that will disappear in the future and that probably affects only a minority of developers please take care of this later.

**Comment by daschuer** [comment]
What is the reason not to use "PC_hidapi_VERSION" that the error message would be reasonable by default.

**Comment by Be-ing** [comment]
That requires rewriting more code to handle the case where there is no pkgconfig.

**Comment by daschuer** [comment]
We have it on all out targets. At least use the "PC_hidapi_VERSION" and fall back to reading the header file, with the code guessing the content of the pc file.

**Comment by Be-ing** [comment]
@daschuer you can improve the error message for outdated operating systems if you want after merging this but I don't think there's any reason not to merge this now. I have a million other things to do.

File: cmake/modules/Findhidapi.cmake (hunk 5 above)
Comment (Holzhaus) [reply cmake/modules/Findhidapi.cmake:65]: Same on Arch:

```
$ yay -Ql hidapi | grep -F hidapi.h
hidapi /usr/include/hidapi/hidapi.h
```

**Comment by daschuer** [comment]
This can instantly merged, if you revert the changes in the CMakeList.txt, and making hidapi_VERSION not a requirement in findhidapi.cmake.
These changes are unnecessary and are leading to the noted issues.

The required bugfix is just to look into the hidapi sub folder, that's all and sufficient for the fix.

**Comment by Be-ing** [comment]
I don't have energy for this bikeshedding.

**Comment by uklotzde** [comment]
@daschuer Instead of starting a discussion it would have been more helpful to simply provide the changes you deem necessary. Then we could test if it still works for us.

**Comment by daschuer** [comment]
@uklotzde It is no problem for me to issue an alternative solution that makes use of the PC_hidapi_VERSION or just a simply fix of the header path. But for my understanding it is the normal way in a review to discuss things with the author. In this case l have  discovered that he original PR was not working at all. Than I have clearly pointed out the later fix introduced an unecessary regression and suggested alternatives. It feels bad to being blamed for bikeshedding because of this. 

Since @Be-ing is stepped back from this let's discuss what would be a merge-able. Should I make use of PC_hidapi_VERSION or is it preferred to just fix the include path based on your original solution?

**Comment by uklotzde** [comment]
@daschuer From what I understood the outcome is expected, i.e. the bundled version is used if the system version is too old and detection fails. The exact error message is a minor detail if it works.

**Comment by uklotzde** [comment]
Why do we need to discuss this if the CI builds demonstrate that there are no actual regressions? This PR fixes the include paths and dynamic linking for all Linux distributions except Ubuntu LTS.

What did I miss?

**Comment by daschuer** [comment]
> The exact error message is a minor detail if it works.

Yes, but we have a solution in the 2.3 branch that has not the wrong error message. This solution was removed here along with the actual fix. I am happy to issue an alternative PR without this regression. I just need a conses before I start to work, because I have no interest to put work into a third solution.

File: CMakeLists.txt [hunk 6]
```diff
@@ -2605,12 +2605,12 @@ find_package(LibUSB)
 # USB HID controller support
 option(HID "USB HID controller support" ON)
 if(HID)
-  find_package(hidapi)
-  # hidapi_VERSION is only available starting with 0.10.0
-  if(NOT hidapi_FOUND OR NOT hidapi_VERSION OR hidapi_VERSION VERSION_LESS 0.10.1)
```
Comment (daschuer) [inline CMakeLists.txt:2610]: This change is unrelated to the fix and can be reverted.

File: cmake/modules/Findhidapi.cmake [hunk 7]
```diff
@@ -62,28 +61,32 @@ find_library(hidapi_LIBRARY
...
-endif ()
+endif()
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(
+  hidapi
+  DEFAULT_MSG
+  hidapi_LIBRARY
+  hidapi_INCLUDE_DIR
+  hidapi_VERSION
```
Comment (daschuer) [inline cmake/modules/Findhidapi.cmake:88]: This makes the hidapi_VERSION version required to "find" the package which is an issue because the code above does not work for versions < 0.10.0. We can make it work by using the version from the PC file or just stick with the original solution.

Comment (daschuer) [reply cmake/modules/Findhidapi.cmake:88]: Please give a hint what you prefer and I will prepare a PR.

**Comment by uklotzde** [comment]
Before: The code was silently broken
After: The code is fixed with slight inconveniences for some special cases

Why do we need to pile up more work on top of it? Changing 1 or 2 lines would be ok, if it is as easy as claimed. But anything else could and should be done later.

**Comment by daschuer** [comment]
I was only asking for reverting a few lines, that's all. 

Let's move one and consider a solution ..

After looking deeper into it I think the Findhidapi.cmake file is still broken.
It aims to work with and without a PC file available, but now the path without PC file will return the wrong include directory
just "/usr/include" instead" of "/usr/include/hidapi". It still woks in our case, because we use the cflags from the PC file, which comes with the correct include derective. 

Do we know why using PATH_SUFFIXES was broken in the first place?

**Comment by daschuer** [comment]
It was not broken, #4215 fixes all issues.

**Comment by Be-ing** [comment]
> just "/usr/include" instead" of "/usr/include/hidapi"

/usr/include is correct. The file is at <hidapi/hidapi.h> relative to the installation prefix.

File: cmake/modules/Findhidapi.cmake (hunk 7 above)
Comment (Be-ing) [reply cmake/modules/Findhidapi.cmake:88]: This is intentional. We require 0.10.1, so if there is no hidapi_VERSION it is too old anyway.

**Comment by Holzhaus** [comment]
> > just "/usr/include" instead" of "/usr/include/hidapi"
> 
> /usr/include is correct. The file is at <hidapi/hidapi.h> relative to the installation prefix.

Nope:

```
$ grep -HrnF hidapi.h src
src/controllers/hid/hidcontroller.cpp:3:#include <hidapi.h>
src/controllers/hid/hidenumerator.cpp:3:#include <hidapi.h>
src/controllers/hid/hiddevice.h:3:#include <hidapi.h>
```

**Comment by Be-ing** [comment]
@Holzhaus yes that is what I mean. The #include paths in our C++ code were wrong. I fixed that in this PR.

**Comment by Holzhaus** [comment]
This is not what the hidapi README suggests: https://github.com/libusb/hidapi#what-does-the-api-look-like

**Comment by Be-ing** [comment]
Well then the documentation is misleading. hidapi's build system has always installed the header to <hidapi/hidapi.h> relative to the installation prefix.

**Comment by Be-ing** [comment]
Fix for the hidapi documentation: https://github.com/libusb/hidapi/pull/323

File: CMakeLists.txt (hunk 6 above)
Comment (Be-ing) [reply CMakeLists.txt:2610]: It is not unrelated. `find_package(hidapi 0.10.1)` creates the error message hinting the version is too old.

**Comment by Be-ing** [comment]
I have already implemented a fully working solution. It does not create a welcoming environment to have every unimportant detail nitpicked to shreds then reimplemented in a worse way that does not solve all the problems.

**Comment by Be-ing** [comment]
The pkgconfig file says the #include path is /usr/include:
```
 ❯ cat /usr/lib64/pkgconfig/hidapi-libusb.pc
prefix=/usr
exec_prefix=/usr
libdir=/usr/lib64
includedir=/usr/include

Name: hidapi-libusb
Description: C Library for USB HID device access from Linux, Mac OS X, FreeBSD, and Windows. This is the libusb implementation.
Version: 0.10.1
Libs: -L${libdir} -lhidapi-libusb
Cflags: -I${includedir}/hidapi
``

#include <hidapi/hidapi.h> is correct.

**Comment by Be-ing** [comment]
Oh, wait the pkgconfig file is doing weird stuff:
```
includedir=/usr/include
```
That is normal.
```
Cflags: -I${includedir}/hidapi
```
That is not.

**Comment by uklotzde** [comment]
The example in `test.c` doesn't even show how hidapi.h is supposed to be included from a system path:

```
#include "hidapi.h"
```

This is NOT how a consumer should include an external library.

**Comment by Holzhaus** [comment]
I think `#include <hidapi.h>` is the standard. Looking at some other prominent projects, most of them use it:
- Yubico's libfido2: https://github.com/Yubico/libfido2/search?l=C&q=hidapi
- Ardour: https://github.com/Ardour/ardour/search?l=C%2B%2B&q=hidapi
- Dolphin Emu: https://github.com/dolphin-emu/dolphin/search?l=C%2B%2B&q=hidapi

File: src/controllers/hid/hidcontroller.h [hunk 8]
```diff
@@ -1,6 +1,6 @@
 #pragma once
 
-#include <hidapi.h>
+#include <hidapi/hidapi.h>
```
Comment (Youw) [inline src/controllers/hid/hidcontroller.h:3]: this is not correct
as per HIDAPI design, you should only include `<hidapi.h>`

**Comment by Be-ing** [comment]
I reverted the change to the #include path.
